apps/scan/backend/src/app_flow.ts

```diff
@@ -0,0 +1,155 @@
+  ) {
+    return 'setup_card_reader';
+  }
+
+  const electionDefinition = store.getElectionDefinition();
+
+  if (
+    !electionDefinition &&
+    authStatus.status === 'logged_out' &&
+    authStatus.reason === 'no_card'
+  ) {
+    return 'login_prompt';
+  }
+
+  if (
+    authStatus.status === 'logged_out' &&
+    authStatus.reason === 'card_error'
+  ) {
+    return 'card_error';
+  }
+
+  if (authStatus.status === 'logged_out' && authStatus.reason !== 'no_card') {
+    return 'invalid_card';
+  }
+
+  if (
+    authStatus.status === 'checking_pin' &&
+    authStatus.user.role === 'system_administrator'
+  ) {
+    return 'unlock_machine';
+  }
+
+  if (isSystemAdministratorAuth(authStatus)) {
+    return 'logged_in:system_administrator';
+  }
+
+  if (precinctScannerState === 'disconnected') {
+    return 'setup_scanner';
+  }
+
+  if (authStatus.status === 'checking_pin') {
+    return 'unlock_machine';
+  }
+
+  if (!electionDefinition) {
+    return 'unconfigured:election';
+  }
+
+  if (isElectionManagerAuth(authStatus)) {
+    return 'logged_in:election_manager';
+  }
+
+  const precinctSelection = store.getPrecinctSelection();
+
+  if (!precinctSelection) {
+    return 'unconfigured:precinct';
+  }
+
+  const usbDriveStatus = await usbDrive.status();
+  if (usbDriveStatus.status !== 'mounted') {
+    return 'insert_usb_drive';
+  }
+
+  const ballotCountWhenBallotBagLastReplaced =
+    store.getBallotCountWhenBallotBagLastReplaced();
+  const ballotsCounted = store.getBallotsCounted();
+  const needsToReplaceBallotBag =
+    ballotsCounted >=
+    ballotCountWhenBallotBagLastReplaced + BALLOT_BAG_CAPACITY;
+  if (needsToReplaceBallotBag && precinctScannerState !== 'accepted') {
+    return 'replace_ballot_bag';
+  }
+
+  if (isPollWorkerAuth(authStatus)) {
+    return 'logged_in:poll_worker';
+  }
+
+  // When no card is inserted, we're in "voter" mode
+  assert(authStatus.status === 'logged_out' && authStatus.reason === 'no_card');
+
+  const pollsState = store.getPollsState();
+  if (pollsState !== 'polls_open') {
+    return 'polls_not_open';
+  }
+
+  if (await doesUsbDriveRequireCastVoteRecordSync(store, usbDriveStatus)) {
+    return 'cast_vote_record_sync_required';
+  }
+
+  switch (precinctScannerState) {
+    case 'accepted':
+      return 'ballot:accepted';
+
+    case 'accepting':
+    case 'accepting_after_review':
+      return 'ballot:accepting';
+
+    case 'scanning':
+      return 'ballot:scanning';
+
+    case 'ready_to_accept':
+      return 'ballot:waiting_to_accept';
+
+    case 'ready_to_scan':
+      return 'ballot:waiting_to_scan';
+
+    case 'connecting':
+    case 'no_paper':
+    case 'returning_to_rescan':
+    case 'needs_review':
+    case 'returning':
+    case 'returned':
+    case 'rejecting':
+    case 'rejected':
+    case 'jammed':
+    case 'both_sides_have_paper':
+    case 'recovering_from_error':
+    case 'double_sheet_jammed':
+    case 'unrecoverable_error':
+      return 'unknown';
```

Comment:
Is there a reason so many of these states (which have meaningful implications for app flow) are mapped to `unknown` here? Is it just that they aren't relevant for deciding if we should scan or not?

Reply:
Yeah, they don't change the decision for whether we should scan or not. This is one of the things that made me think this whole thing is a little overwrought and it should just determine whether we're ready to scan or not. I'd like it to be the thing that controls the frontend state in the future, but we're not really there yet. Thoughts?

Reply:
I think it's interesting to think about having the entire app flow state controlled by the backend. On reading this PR, I started thinking harder about what that would be like. It does feel like a big jump (and would seriously set VxScan apart from the other apps). I considered whether we should then move all of that logic into the state machine.

I do think a smaller change where we just have a guard on the backend that checks all of the conditions we care about before initiating a scan seems less intimidating for now. I almost suggested it, but wanted to understand where you were going with this first. Given that you are feeling a similar hesitation, I think we might at least want to explore that option.

Reply:
Simplified this a bit in 69fcb1be.